### PR TITLE
[SYM-115] Support border width, color and corner radius for PAN

### PR DESCRIPTION
**Pull request by milosbswe**
## What

Add support for border width, border color and corner radius for PAN field.

## Test Plan

- ❌ I've added unit tests for this change.
- ❌ The reviewer should manually test the changes in this PR.

## Demo

Focused:

<table><tr>
<td> 
  <p align="center" style="padding: 10px">
Screenshot 2023-09-12 at 16 17 31
    <br>
    <em style="color: grey">Default</em>
  </p> 
</td>
<td> 
  <p align="center">
Screenshot 2023-09-12 at 16 16 26
    <br>
    <em style="color: grey">Custom</em>
  </p> 
</td>
</tr></table>

Unfocused:

<table><tr>
<td> 
  <p align="center" style="padding: 10px">
Screenshot 2023-09-12 at 16 17 44
    <br>
    <em style="color: grey">Default</em>
  </p> 
</td>
<td> 
  <p align="center">
Screenshot 2023-09-12 at 16 35 56
    <br>
    <em style="color: grey">Custom</em>
  </p> 
</td>
</tr></table>

### CODE REVIEW

File: forage-android/src/main/java/com/joinforage/forage/android/ui/ForageElement.kt [hunk 1]
```diff
@@ -21,6 +21,7 @@ interface ForageElement {
     fun setTextSize(textSize: Float)
     fun setHint(hint: String)
     fun setHintTextColor(hintTextColor: Int)
+    fun setBoxStrokeColor(boxStrokeColor: Int)
```
Comment (devinmorgan) [inline forage-android/src/main/java/com/joinforage/forage/android/ui/ForageElement.kt:24]: nice! Thank you for this

File: forage-android/src/main/java/com/joinforage/forage/android/ui/FormatPanTextWatcher.kt [hunk 2]
```diff
@@ -79,6 +82,36 @@ fun getTransformedPosition(pos: Int, oldPrettyText: String, newPrettyText: Strin
...
+        getDimension(com.joinforage.forage.android.R.styleable.ForagePANEditText_boxCornerBottomEnd, 0f)
+    return if (boxCornerRadiusBottomEnd == 0f) boxCornerRadius else boxCornerRadiusBottomEnd
+}
+
+fun TypedArray.getBoxCornerRadiusTopStart(boxCornerRadius: Float): Float {
+    val boxCornerRadiusTopStart =
+        getDimension(com.joinforage.forage.android.R.styleable.ForagePANEditText_boxCornerTopStart, 0f)
+    return if (boxCornerRadiusTopStart == 0f) boxCornerRadius else boxCornerRadiusTopStart
+}
+
```
Comment (devinmorgan) [inline forage-android/src/main/java/com/joinforage/forage/android/ui/FormatPanTextWatcher.kt:114]: Are these related to the FormatPanTextWatcher? If not, can we create a new file for these methods?

File: forage-android/src/main/res/values/attrs.xml [hunk 3]
```diff
@@ -12,6 +12,16 @@
...
+
+        <attr name="cornerRadius" />
+        <attr format="dimension" name="boxCornerTopEnd" />
+        <attr format="dimension" name="boxCornerTopStart" />
+        <attr format="dimension" name="boxCornerBottomStart" />
+        <attr format="dimension" name="boxCornerBottomEnd" />
+
+        <attr name="android:strokeColor" />
+        <attr name="android:strokeWidth" />
+        <attr name="boxStrokeWidthFocused" />
```
Comment (devinmorgan) [inline forage-android/src/main/res/values/attrs.xml:24]: The `boxCorner*`s all of `format=`s fields. Any particular reason we don't want the `format=` on the other `<attr />`s we're introducing in this PR?

File: forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePANEditText.kt [hunk 4]
```diff
@@ -55,17 +56,54 @@ class ForagePANEditText @JvmOverloads constructor(
...
+                    val boxCornerRadiusBottomStart = getBoxCornerRadiusBottomStart(focusedBoxCornerRadius)
+                    val boxCornerRadiusBottomEnd = getBoxCornerRadiusBottomEnd(focusedBoxCornerRadius)
+
                     textInputLayout =
                         TextInputLayout(context, null, textInputLayoutStyleAttribute).apply {
                             layoutParams =
                                 LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
+
+                            boxStrokeColor = focusedBoxStrokeColor
+                            boxStrokeWidthFocused = focusedBoxStrokeWidth.toInt()
```
Comment (devinmorgan) [inline forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePANEditText.kt:96]: I'm not very familiar with a lot of Android styling. How come we have a `focusedBoxStrokeWidth` and `strokeWidth` but we have only `focusedBoxStrokeColor` and no `strokeColor`? Asking so I can learn!

**Review by devinmorgan — COMMENTED** [review]
Nice work so far! A couple of questions I was hoping to get clarity on

File: forage-android/src/main/java/com/joinforage/forage/android/ui/FormatPanTextWatcher.kt (hunk 2 above)
Comment (djoksimo) [reply forage-android/src/main/java/com/joinforage/forage/android/ui/FormatPanTextWatcher.kt:114]: Can we also make sure that internal-facing code uses the `internal`  access modifier. Currently, these methods may default to `public`

**Review by djoksimo — CHANGES_REQUESTED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/ui/ForageElement.kt (hunk 1 above)
Comment (milosbswe) [reply forage-android/src/main/java/com/joinforage/forage/android/ui/ForageElement.kt:24]: Thanks!

File: forage-android/src/main/java/com/joinforage/forage/android/ui/FormatPanTextWatcher.kt (hunk 2 above)
Comment (milosbswe) [reply forage-android/src/main/java/com/joinforage/forage/android/ui/FormatPanTextWatcher.kt]: Updated. Thanks!

File: forage-android/src/main/res/values/attrs.xml (hunk 3 above)
Comment (milosbswe) [reply forage-android/src/main/res/values/attrs.xml]: I added "format" fields to have consistency. Thanks!

File: forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePANEditText.kt (hunk 4 above)
Comment (milosbswe) [reply forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePANEditText.kt:97]: Good question! I think(based on my research) that Android TextInputLayout does not provide a direct method to set the stroke color programmatically. The stroke color inside the TextInputLayout is typically controlled through XML styles. 
So, stroke width and color, and unfocused stroke width can be set programmatically. Unfocused stroke color can be changed through `styles.xml`. 
Also, I added one-line comments above those setters for better understanding what we have and what we can change through ForagePANEditText.

Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePANEditText.kt]: really solid explanation - appreciate it a lot!

File: forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePanTextUtils.kt
Comment (devinmorgan) [inline forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePanTextUtils.kt:1]: Nice! Appreciate you moving these (and using the `internal`) modifier

**Review by devinmorgan — APPROVED** [review]
Thanks for the changes, Milos! Let's wait for @djoksimo to give the thumbs up as well before merging since he also requested some changes. Nice work!

**Review by djoksimo — APPROVED** [review]
